### PR TITLE
Javascript javascript1 week2

### PR DESCRIPTION
**Pull request by bhumikama**
Implemented below tasks in each of the commits 
1.Flight booking fullname function
2.Event Application
3.Weather Wear
4.Student Manager
5.Candy Helper

### CODE REVIEW

File: javascript/javascript1/week2/answer.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+function getFullname(firstname, surname, useFormalName, isWoman) {
+  if (useFormalName) {
+    if (isWoman) {
+      return "Madam " + firstname + " " + surname;
+    } else {
+      return "Lord " + firstname + " " + surname;
+    }
+  } else {
+    return firstname + " " + surname;
+  }
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/answer.js]: You could safe writing a few extra characters by writing only this:
```suggestion
  }
  
return firstname + " " + surname;
```

File: javascript/javascript1/week2/answer.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+const fullname1 = "Benjamin";
+const fullname2 = "Hughes";
+const useFormal = true;
+const isFemale = true;
+const display = getFullname(fullname1, fullname2, useFormal, isFemale);
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/answer.js:23]: If for some reason there is no argument passed for `isFemale` the code will execute with "Lord Benjamin Hughes", while it might be better to fall into the case of simply "Benjamin Hughes". You don't need to fix this, but just wanted to bring your attention to testing for edge cases like this.

File: javascript/javascript1/week2/answer.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
+//Flight booking fullname function
+/*
+this function returns a string concatenating
+firstname and surname and also checks if the name needs to be addressed formally,
+if so then checks if the person is a woman and returns the string accordingly
+*/
+function getFullname(firstname, surname, useFormalName, isWoman) {
+  if (useFormalName) {
+    if (isWoman) {
+      return "Madam " + firstname + " " + surname;
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/answer.js]: There is a coding principle called [DRY (Don't Repeat Yourself)](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) which recommends creating functions and variables for each instance where you could repeat a line or block of code more than once. In this case, you use `firstname + " " + surname` 3 times. I'd recommend you setting it once as a variable outside of the `if else` statements and then using it as a variable. eg:
```
const fullName = firstname + " " + surname;
```
and further down as `return "Madam " + fullName;`

File: javascript/javascript1/week2/answer.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+const useFormal = true;
+const isFemale = true;
+const display = getFullname(fullname1, fullname2, useFormal, isFemale);
+console.log(display);
+
+//Event application
+function getEventWeekday(daysFromToday) {
+  let dateObject = new Date();
+  let dayNumber = dateObject.getDay(); //returns a number from 0-6
+  let dayNames = [
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/answer.js]: You did great in the previous exercise using `const` but here it's the opposite. None of the variables you've set as `let` actually is reassigned, and in that case you can easily set them all to `const`.

File: javascript/javascript1/week2/answer.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    "Monday",
+    "Tuesday",
+    "Wednesday",
+    "Thursday",
+    "Friday",
+    "Saturday",
+  ];
+  const indexes = daysFromToday + dayNumber;
+  if (indexes >= 7) {
+    let k = indexes % 7; //since dayNames has only 7 elements need to find the remainder
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/answer.js]: `k` could be a `const`, but further it could also benefit from a more descriptive variable name. Yet, good practice with the code comment on the side.

File: javascript/javascript1/week2/answer.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    return "multiple layers of coat and jacket";
+  } else if (temperature >= 0 && temperature <= 10) {
+    return "sweaters and long pants";
+  } else if (temperature > 10 && temperature < 18) {
+    return "light coats or cotton clothes";
+  } else {
+    return "shorts and a t-shirt";
+  }
+}
+const clothesToWear = findTypeOfClothes(18);
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/answer.js:65]: Well done. This exercise is perfect.
If you would like to practice more, you could check in the beginning of the function if `type of temperature` is equal to number, and if it isn't to return some error message.

File: javascript/javascript1/week2/answer.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+console.log(`You can wear ${clothesToWear}`);
+
+//Student Manager
+const class07Students = [];
+function addStudentToClass(studentName) {
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/answer.js:70]: Beautifully done!

File: javascript/javascript1/week2/answer.js [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+      return "Madam " + firstname + " " + surname;
+    } else {
+      return "Lord " + firstname + " " + surname;
+    }
+  } else {
+    return firstname + " " + surname;
+  }
+}
+
+const fullname1 = "Benjamin";
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/answer.js:19]: I know this is how the variable is written in the exercise requirements, but it would be better to be in camelCase. The same is valid for the function name.

File: javascript/javascript1/week2/answer.js [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+      class07Students.push(studentName);
+    }
+  } else {
+    console.log("Cannot add more students to class 07");
+  }
+}
+
+function getNumberOfStudents() {
+  return class07Students.length;
+}
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/answer.js:84]: Generally, we could just directly use `class07Students.length` instead of creating a function for it, but again the exercise is explicit about doing it with a function.

File: javascript/javascript1/week2/answer.js [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+  },
+];
+
+const boughtCandyPrices = [];
+let totalAmount = 0; //total amount spent on buying candies
+const amountToSpend = Math.floor(Math.random() * 100); // to get a whole number
+console.log(`${amountToSpend}`);
+
+function addCandy(candyType, weight) {
+  let candyItem = candies.find((candy) => candy.type === candyType); //returns the first element from candies array that meets the condition
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/answer.js]: `const`

File: javascript/javascript1/week2/answer.js [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    type: "Toffee",
+    pricePerGram: 1.1,
+  },
+  {
+    type: "Chewing-gum",
+    pricePerGram: 0.03,
+  },
+];
+
+const boughtCandyPrices = [];
```
Comment (jason-vasilev) [inline javascript/javascript1/week2/answer.js:122]: Thank you for doing this optional exercise. It's the hardest for that week and you've done great. Congratulations on the effort and dedication!

**Review by jason-vasilev — APPROVED** [review]
Bhumika, you've done perfectly well. I've written a few comments, but it's mostly suggestions about doing things differently, else your code works as it's supposed to. You are welcome to practice more or merge.

File: javascript/javascript1/week2/answer.js (hunk 2 above)
Comment (bhumikama) [reply javascript/javascript1/week2/answer.js:23]: I used that logic thinking if the user wants to be addressed using formal name, then for female user it is going to concatenate fullName with "Madam" and for male users it is going to use "Lord" instead.Please suggest if this is correct.

**Comment by bhumikama** [comment]
Thanks a lot for providing detailed explanations for each of the task and for providing all the suggestions. I have tried to make changes as suggested.

File: javascript/javascript1/week2/answer.js (hunk 2 above)
Comment (jason-vasilev) [reply javascript/javascript1/week2/answer.js:23]: Your logic is correct. I was inclining more on the fact that you won't always get perfect data. Imagine someone simply calls the function as `getFullname("Josephine", "Jensen", true);`. What should the code do in this case? The last argument for gender is missing. It can be wrong to make an assumption for either male or female and set one as default. The best would be to skip the formal part, if `useFormal` is `true`, but there is no argument provided for `isFemale`.

**Review by jason-vasilev — APPROVED** [review]
